File: crates/deepsec-cli/src/commands/export.rs
use crate::config_loader::Context;
use anyhow::Result;
use clap::Args as ClapArgs;
use deepsec_core::{RevalidationVerdict, Severity, store::load_all_file_records};

#[derive(Debug, ClapArgs)]
pub struct Args {
    #[arg(long)]
    pub project_id: String,
    /// Minimum severity to include.
    #[arg(long)]
    pub min_severity: Option<String>,
    /// Only export findings produced by this run.
    #[arg(long)]
    pub run_id: Option<String>,
    /// Verdict filter (true-positive, false-positive, fixed, uncertain,
    /// accepted-risk).
    #[arg(long)]
    pub verdict: Option<String>,
    /// Only this vulnerability slug.
    #[arg(long)]
    pub slug: Option<String>,
    /// File-path prefix filter.
    #[arg(long)]
    pub prefix: Option<String>,
    /// Write JSON to this path instead of stdout.
    #[arg(long)]
    pub output: Option<std::path::PathBuf>,
}

pub fn run(args: Args, ctx: &Context) -> Result<()> {
    let records = load_all_file_records(&ctx.data_root, &args.project_id)?;
    let min_sev = args
        .min_severity
        .as_deref()
        .map(parse_severity)
        .transpose()?
        .unwrap_or(Severity::Low);
    let want_verdict = args.verdict.as_deref().map(parse_verdict).transpose()?;

    let mut rows = Vec::new();
    for rec in &records {
        if let Some(p) = &args.prefix {
            if !rec.file_path.starts_with(p) {
                continue;
            }
        }
        for f in &rec.findings {
            if f.severity.rank() < min_sev.rank() {
                continue;
            }
            if let Some(rid) = &args.run_id {
                if f.produced_by_run_id.as_deref() != Some(rid.as_str()) {
                    continue;
                }
            }
            if let Some(slug) = &args.slug {
                if f.vuln_slug != *slug {
                    continue;
                }
            }
            if let Some(v) = want_verdict {
                let actual = f.revalidation.as_ref().map(|r| r.verdict);
                if actual != Some(v) {
                    continue;
                }
            }
            rows.push(serde_json::json!({
                "filePath": rec.file_path,
                "finding": f,
                "gitInfo": rec.git_info,
            }));
        }
    }

    let body = serde_json::to_string_pretty(&rows)?;
    if let Some(out) = args.output {
        if let Some(parent) = out.parent() {
            fs_err::create_dir_all(parent)?;
        }
        fs_err::write(&out, &body)?;
        eprintln!("wrote {} findings to {}", rows.len(), out.display());
    } else {
        println!("{body}");
    }
    Ok(())
}

fn parse_severity(s: &str) -> Result<Severity> {
    Ok(match s.to_uppercase().as_str() {
        "CRITICAL" => Severity::Critical,
        "HIGH" => Severity::High,
        "HIGH_BUG" => Severity::HighBug,
        "MEDIUM" => Severity::Medium,
        "BUG" => Severity::Bug,
        "LOW" => Severity::Low,
        _ => anyhow::bail!("unknown severity: {s}"),
    })
}

fn parse_verdict(s: &str) -> Result<RevalidationVerdict> {
    Ok(match s.to_lowercase().as_str() {
        "true-positive" | "tp" => RevalidationVerdict::TruePositive,
        "false-positive" | "fp" => RevalidationVerdict::FalsePositive,
        "fixed" => RevalidationVerdict::Fixed,
        "uncertain" => RevalidationVerdict::Uncertain,
        "accepted-risk" => RevalidationVerdict::AcceptedRisk,
        _ => anyhow::bail!("unknown verdict: {s}"),
    })
}

File: internal/cli/commands/export.go
package commands

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/noeljackson/deepsec/internal/cli"
	"github.com/noeljackson/deepsec/internal/core"
	"github.com/spf13/cobra"
)

// NewExportCmd is a filtered JSON dump of findings.
func NewExportCmd(loader func() (*cli.Context, error)) *cobra.Command {
	var projectID, minSeverity, runID, verdict, slug, prefix, output, format string
	cmd := &cobra.Command{
		Use:   "export",
		Short: "Filtered JSON export of findings",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ctx, err := loader()
			if err != nil {
				return err
			}
			records, err := ctx.DataRoot.LoadAllFileRecords(projectID)
			if err != nil {
				return err
			}
			minSev := core.SeverityLow
			if minSeverity != "" {
				minSev = core.Severity(strings.ToUpper(minSeverity))
				if minSev.Rank() == 0 {
					return fmt.Errorf("unknown severity: %s", minSeverity)
				}
			}
			var wantVerdict *core.RevalidationVerdict
			if verdict != "" {
				v := core.RevalidationVerdict(strings.ToLower(verdict))
				wantVerdict = &v
			}

			out := make([]exportRow, 0)
			for _, rec := range records {
				if prefix != "" && !strings.HasPrefix(rec.FilePath, prefix) {
					continue
				}
				for i := range rec.Findings {
					f := &rec.Findings[i]
					if f.Severity.Rank() < minSev.Rank() {
						continue
					}
					if runID != "" && f.ProducedByRunID != runID {
						continue
					}
					if slug != "" && f.VulnSlug != slug {
						continue
					}
					if wantVerdict != nil {
						if f.Revalidation == nil || f.Revalidation.Verdict != *wantVerdict {
							continue
						}
					}
					out = append(out, exportRow{FilePath: rec.FilePath, Finding: f, GitInfo: rec.GitInfo})
				}
			}
			switch strings.ToLower(format) {
			case "", "json":
				body, err := json.MarshalIndent(out, "", "  ")
				if err != nil {
					return err
				}
				return write(output, body)
			case "sarif":
				body, err := renderSARIF(out, projectID)
				if err != nil {
					return err
				}
				return write(output, body)
			}
			return fmt.Errorf("unknown format %q (try 'json' or 'sarif')", format)
		},
	}
	cmd.Flags().StringVar(&projectID, "project-id", "", "Project id (required)")
	_ = cmd.MarkFlagRequired("project-id")
	cmd.Flags().StringVar(&minSeverity, "min-severity", "", "Minimum severity")
	cmd.Flags().StringVar(&runID, "run-id", "", "Only findings produced by this run")
	cmd.Flags().StringVar(&verdict, "verdict", "", "true-positive / false-positive / fixed / uncertain / accepted-risk")
	cmd.Flags().StringVar(&slug, "slug", "", "Only this vulnSlug")
	cmd.Flags().StringVar(&prefix, "prefix", "", "File-path prefix filter")
	cmd.Flags().StringVar(&output, "output", "", "Write to this path (default: stdout)")
	cmd.Flags().StringVar(&format, "format", "json", "json | sarif")
	return cmd
}

func write(path string, body []byte) error {
	if path == "" {
		fmt.Println(string(body))
		return nil
	}
	return writePrivateFile(path, body)
}

// exportRow is the on-wire shape for JSON export. SARIF emission
// reuses the same rows.
type exportRow struct {
	FilePath string        `json:"filePath"`
	Finding  *core.Finding `json:"finding"`
	GitInfo  *core.GitInfo `json:"gitInfo,omitempty"`
}

// renderSARIF produces a minimal SARIF 2.1.0 document. Covers the
// fields GitHub Code Scanning consumes; deeper SARIF features (taxa,
// related locations, etc.) can be added if needed.
func renderSARIF(rows []exportRow, projectID string) ([]byte, error) {
	results := make([]map[string]any, 0, len(rows))
	rules := map[string]map[string]any{}

	for _, r := range rows {
		ruleID := r.Finding.VulnSlug
		if _, ok := rules[ruleID]; !ok {
			rules[ruleID] = map[string]any{
				"id":   ruleID,
				"name": ruleID,
				"shortDescription": map[string]any{
					"text": r.Finding.Title,
				},
				"helpUri": "https://github.com/noeljackson/deepsec",
			}
		}
		startLine := 1
		if len(r.Finding.LineNumbers) > 0 {
			startLine = r.Finding.LineNumbers[0]
		}
		results = append(results, map[string]any{
			"ruleId":  ruleID,
			"level":   sarifLevel(r.Finding.Severity),
			"message": map[string]any{"text": r.Finding.Description},
			"locations": []map[string]any{{
				"physicalLocation": map[string]any{
					"artifactLocation": map[string]any{"uri": r.FilePath},
					"region":           map[string]any{"startLine": startLine},
				},
			}},
		})
	}

	ruleList := make([]map[string]any, 0, len(rules))
	for _, v := range rules {
		ruleList = append(ruleList, v)
	}

	doc := map[string]any{
		"$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
		"version": "2.1.0",
		"runs": []map[string]any{{
			"tool": map[string]any{
				"driver": map[string]any{
					"name":    "deepsec",
					"rules":   ruleList,
					"version": "0.1.0",
				},
			},
			"results": results,
		}},
	}
	return json.MarshalIndent(doc, "", "  ")
}

func sarifLevel(s core.Severity) string {
	switch s {
	case core.SeverityCritical, core.SeverityHigh, core.SeverityHighBug:
		return "error"
	case core.SeverityMedium, core.SeverityBug:
		return "warning"
	}
	return "note"
}
